services/QuillLMS/app/queries/admin_diagnostic_reports/diagnostic_aggregate_query.rb

```diff
@@ -0,0 +1,180 @@
+
+      @additional_aggregation = aggregation
+
+      super(**options)
+    end
+
+    def run
+      post_process(run_query)
+    end
+
+    def select_clause
+      <<-SQL
+        SELECT
+          activities.id AS diagnostic_id,
+          activities.name AS diagnostic_name,
+          #{aggregate_by_clause} AS aggregate_id,
+          #{aggregate_sort_clause} AS name,
+          #{specific_select_clause}
+      SQL
+    end
+
+    def from_and_join_clauses
+      super + <<-SQL
+        JOIN lms.classroom_units
+          ON classrooms.id = classroom_units.classroom_id
+        JOIN lms.users
+          ON schools_users.user_id = users.id
+      SQL
+    end
+
+    def specific_select_clause
+      raise NotImplementedError
+    end
+
+    def where_clause
+      super + <<-SQL
+          #{relevant_diagnostic_where_clause}
+      SQL
+    end
+
+    def relevant_diagnostic_where_clause
+      "AND activities.id IN (#{DIAGNOSTIC_ORDER_BY_ID.join(',')})"
+    end
+
+    def group_by_clause
+      "GROUP BY activities.id, activities.name, aggregate_id, #{aggregate_sort_clause}"
+    end
+
+    def aggregate_by_clause
+      {
+        'grade' => "classrooms.grade",
+        'classroom' => "classrooms.id",
+        'teacher' => "users.id"
+      }.fetch(additional_aggregation)
+    end
+
+    def aggregate_sort_clause
+      {
+        'grade' => "classrooms.grade",
+        'classroom' => "classrooms.name",
+        'teacher' => "users.name"
+      }.fetch(additional_aggregation)
+    end
+
+    private def post_process(result)
+      return [] if result.empty?
+
+      result.group_by { |row| row[:diagnostic_name] }
+        .values
+        .map { |diagnostic_rows| build_diagnostic_aggregates(diagnostic_rows) }
+        .sort_by { |diagnostic| DIAGNOSTIC_ORDER_BY_ID.index(diagnostic[:diagnostic_id]) }
+    end
+
+    private def build_diagnostic_aggregates(diagnostic_rows)
+      {
+        id: diagnostic_rows.first[:diagnostic_id],
+        name: diagnostic_rows.first[:diagnostic_name],
+        group_by: additional_aggregation,
+        aggregate_rows: process_aggregate_rows(diagnostic_rows)
+      }.merge(aggregate_diagnostic(diagnostic_rows))
+    end
+
+    private def process_aggregate_rows(diagnostic_rows)
+      aggregate_sort(diagnostic_rows)
+        .map do |row|
+          # we only need aggregate_id during the SQL grouping, and don't need to pass that value on
+          row.delete(:aggregate_id)
+
+          # Make grade information more human-readable than simple integers
+          next row unless grade_aggregation?
+          next row.merge({name: "No grade selected"}) if row[:name].nil?
+          # to_i returns 0 for non-numeric strings, so this will only apply to numbered grades
+          next row.merge({name: "Grade #{row[:name]}"}) if row[:name].to_i > 0
+
+          row
+        end
+    end
+
+    private def aggregate_sort(diagnostic_rows)
+      diagnostic_rows.sort do |a, b|
```

Comment:
Purely a preference thing, but it might be easier to run these sorts using [`sort_by`](https://ruby-doc.org/core-3.0.2/Enumerable.html#method-i-sort_by), since the syntax is a little easier to parse, e.g. 

```ruby
diagnostic_rows.sort_by do |row|
  ## TODO grade logic
  ## TODO teacher logic
  row[:name]
end